File: packages/go/types/property_test.go
package types

import (
	"testing"
	"testing/quick"

	"pgregory.net/rapid"
)

func TestPropertyPhoneNumberValidation(t *testing.T) {
	check := func(phone string) bool {
		valid := IsPhoneNumberValid(phone)
		if valid {
			if len(phone) != 12 {
				return false
			}
			if phone[:4] != "2547" {
				return false
			}
		}
		return true
	}
	if err := quick.Check(check, nil); err != nil {
		t.Error(err)
	}
}

func TestPropertyFormatPhoneNumber(t *testing.T) {
	rapid.Check(t, func(t *rapid.T) {
		phone := rapid.StringMatching("^0?7[0-9]{8}$").Draw(t, "phone")
		formatted := FormatPhoneNumber(phone)
		if len(formatted) != 12 {
			t.Fatalf("expected 12 chars, got %d: %s", len(formatted), formatted)
		}
		if formatted[:4] != "2547" {
			t.Fatalf("expected 2547 prefix, got %s", formatted[:4])
		}
	})
}

func TestPropertyMaskSensitiveData(t *testing.T) {
	rapid.Check(t, func(t *rapid.T) {
		raw := rapid.MapOf(
			rapid.SampledFrom([]string{"consumerKey", "Password", "SecurityCredential", "otherField"}),
			rapid.String(),
		).Draw(t, "data")

		data := make(map[string]interface{}, len(raw))
		for k, v := range raw {
			data[k] = v
		}

		masked := MaskSensitiveData(data)
		for _, key := range []string{"consumerKey", "Password", "SecurityCredential"} {
			if v, ok := masked[key]; ok {
				val := v.(string)
				if len(data[key].(string)) > 0 {
					if val[len(val)-4:] != "****" {
						t.Fatalf("expected masked value to end with ****, got %s", val)
					}
				}
			}
		}
	})
}

func TestPropertyIdempotencyKeyDeterministic(t *testing.T) {
	rapid.Check(t, func(t *rapid.T) {
		method := rapid.StringN(1, 5, 10).Draw(t, "method")
		url := rapid.StringN(1, 10, 50).Draw(t, "url")
		body := rapid.MapOf(
			rapid.String(),
			rapid.Int(),
		).Draw(t, "body")

		key1 := GenerateIdempotencyKey(method, url, body)
		key2 := GenerateIdempotencyKey(method, url, body)
		if key1 != key2 {
			t.Fatalf("expected identical keys for same inputs, got %s != %s", key1, key2)
		}
	})
}

func TestPropertyIdempotencyKeyDifferent(t *testing.T) {
	rapid.Check(t, func(t *rapid.T) {
		m1 := rapid.StringN(1, 3, 5).Draw(t, "m1")
		u1 := rapid.StringN(1, 5, 10).Draw(t, "u1")
		m2 := rapid.StringN(1, 3, 5).Draw(t, "m2")
		u2 := rapid.StringN(1, 5, 10).Draw(t, "u2")

		if m1 == m2 && u1 == u2 {
			return
		}
		key1 := GenerateIdempotencyKey(m1, u1, map[string]interface{}{})
		key2 := GenerateIdempotencyKey(m2, u2, map[string]interface{}{})
		if key1 == key2 {
			t.Fatalf("expected different keys for different inputs, got %s", key1)
		}
	})
}

func TestPropertyNormalizeEndpointKey(t *testing.T) {
	rapid.Check(t, func(t *rapid.T) {
		key := rapid.OneOf(
			rapid.StringMatching("^https://[a-z]+\\.[a-z]+/[a-z]+/[a-z]+$"),
			rapid.StringMatching("^[A-Z]+/[A-Z]+/[A-Z]+$"),
		).Draw(t, "key")

		normalized := normalizeEndpointKey(key)
		if len(normalized) == 0 {
			t.Fatal("expected non-empty normalized key")
		}
	})
}

File: sdks/typescript/tests/unit/property.test.ts
import { describe, it, expect } from "vitest";
import fc from "fast-check";
import {
  isPhoneNumberValid,
  formatPhoneNumber,
  maskSensitiveData,
} from "../../src/utils/index.js";
import { normalizeEndpointKey } from "../../src/utils/rate-limiter.js";
import { generateIdempotencyKey } from "../../src/utils/idempotency.js";

describe("Property-based: Phone Number Validation", () => {
  it("should reject all non-2547XXXXXXXX numbers", () => {
    fc.assert(
      fc.property(fc.string(), (phone) => {
        fc.pre(phone.length > 0);
        const valid = isPhoneNumberValid(phone);
        const startsWith2547 = phone.startsWith("2547");
        const has8Digits = /^2547\d{8}$/.test(phone);
        return !valid || (startsWith2547 && has8Digits);
      }),
      { numRuns: 1000 },
    );
  });

  it("formatPhoneNumber should always produce a valid number", () => {
    fc.assert(
      fc.property(
        fc.oneof(
          fc.integer({ min: 700000000, max: 799999999 }).map(String),
          fc.stringMatching(/^0?7\d{8}$/),
          fc.stringMatching(/^2547\d{8}$/),
        ),
        (phone) => {
          fc.pre(phone.length > 0);
          const formatted = formatPhoneNumber(phone);
          return /^2547\d{8}$/.test(formatted);
        },
      ),
      { numRuns: 1000 },
    );
  });
});

describe("Property-based: Masking", () => {
  it("should never expose full sensitive values", () => {
    fc.assert(
      fc.property(
        fc.record({
          consumerKey: fc.string({ minLength: 1, maxLength: 30 }),
          Password: fc.string({ minLength: 1, maxLength: 30 }),
          otherField: fc.string({ minLength: 1, maxLength: 30 }),
        }),
        (data) => {
          const masked = maskSensitiveData(data);
          if ("consumerKey" in masked) {
            expect(String(masked.consumerKey)).toContain("****");
          }
          if ("Password" in masked) {
            expect(String(masked.Password)).toContain("****");
          }
          expect(masked.otherField).toBe(data.otherField);
        },
      ),
      { numRuns: 500 },
    );
  });
});

describe("Property-based: Endpoint Key Normalization", () => {
  it("should always produce lowercase strings without protocol", () => {
    fc.assert(
      fc.property(
        fc.oneof(
          fc.stringMatching(/^https:\/\/[a-z]+\.[a-z]+\/[a-z]+\/[a-z]+$/),
          fc.stringMatching(/^[A-Z]+\/[A-Z]+\/[A-Z]+$/),
        ),
        (key) => {
          const normalized = normalizeEndpointKey(key);
          expect(normalized).not.toContain("https://");
          expect(normalized).not.toContain("http://");
          expect(normalized).toEqual(normalized.toLowerCase());
        },
      ),
      { numRuns: 500 },
    );
  });
});

describe("Property-based: Idempotency Key", () => {
  it("should be deterministic for same inputs", () => {
    fc.assert(
      fc.property(
        fc.string({ minLength: 3, maxLength: 10 }),
        fc.string({ minLength: 5, maxLength: 50 }),
        fc.record({ a: fc.integer(), b: fc.string() }),
        (method, url, body) => {
          const key1 = generateIdempotencyKey(method, url, body);
          const key2 = generateIdempotencyKey(method, url, body);
          expect(key1).toBe(key2);
        },
      ),
      { numRuns: 500 },
    );
  });

  it("should produce different keys for different inputs", () => {
    fc.assert(
      fc.property(
        fc.string({ minLength: 3, maxLength: 5 }),
        fc.string({ minLength: 5, maxLength: 10 }),
        fc.string({ minLength: 3, maxLength: 5 }),
        fc.string({ minLength: 5, maxLength: 10 }),
        (m1, u1, m2, u2) => {
          fc.pre(m1 !== m2 || u1 !== u2);
          const key1 = generateIdempotencyKey(m1, u1, {});
          const key2 = generateIdempotencyKey(m2, u2, {});
          expect(key1).not.toBe(key2);
        },
      ),
      { numRuns: 500 },
    );
  });
});
